## Objectives: clean for nulls and split data into three time periods: 
- Jan 2020 - April 2020
- Jan 2020 - Aug 2020
- Jan 2020 - Dec 2020

In [ ]:
import pandas as pd

In [3]:
df = pd.read_csv('C:/Users/tlan4/Documents/Practice Projects/POC-CSV-Pivot-Plots/Topic_Input/WeatherEvents_Jan2016-Dec2020.csv')

In [56]:
df.head()

EventId  Type Severity      StartTime(UTC)        EndTime(UTC)     TimeZone  \
0     W-1  Snow    Light 2016-01-06 23:14:00 2016-01-07 00:34:00  US/Mountain   
1     W-2  Snow    Light 2016-01-07 04:14:00 2016-01-07 04:54:00  US/Mountain   
2     W-3  Snow    Light 2016-01-07 05:54:00 2016-01-07 15:34:00  US/Mountain   
3     W-4  Snow    Light 2016-01-08 05:34:00 2016-01-08 05:54:00  US/Mountain   
4     W-5  Snow    Light 2016-01-08 13:54:00 2016-01-08 15:54:00  US/Mountain   

  AirportCode  LocationLat  LocationLng      City    County State  ZipCode  
0        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
1        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
2        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
3        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
4        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6274206 entries, 0 to 6274205
Data columns (total 13 columns):
 #   Column          Dtype  
---  ------          -----  
 0   EventId         object 
 1   Type            object 
 2   Severity        object 
 3   StartTime(UTC)  object 
 4   EndTime(UTC)    object 
 5   TimeZone        object 
 6   AirportCode     object 
 7   LocationLat     float64
 8   LocationLng     float64
 9   City            object 
 10  County          object 
 11  State           object 
 12  ZipCode         float64
dtypes: float64(3), object(10)
memory usage: 622.3+ MB


In [11]:
df.shape

(6274206, 13)

In [12]:
df.columns

Index(['EventId', 'Type', 'Severity', 'StartTime(UTC)', 'EndTime(UTC)',
       'TimeZone', 'AirportCode', 'LocationLat', 'LocationLng', 'City',
       'County', 'State', 'ZipCode'],
      dtype='object')

In [13]:
df.describe()

LocationLat   LocationLng       ZipCode
count  6.274206e+06  6.274206e+06  6.225569e+06
mean   3.875908e+01 -9.191729e+01  5.251220e+04
std    5.469350e+00  1.342327e+01  2.563980e+04
min    2.455710e+01 -1.245550e+02  1.022000e+03
25%    3.455750e+01 -9.782830e+01  3.140900e+04
50%    3.932000e+01 -8.983690e+01  5.416600e+04
75%    4.301940e+01 -8.195650e+01  7.350500e+04
max    4.894020e+01 -6.779280e+01  9.936200e+04

#### Check for Number of Nulls

In [14]:
df.isnull().sum()

EventId               0
Type                  0
Severity              0
StartTime(UTC)        0
EndTime(UTC)          0
TimeZone              0
AirportCode           0
LocationLat           0
LocationLng           0
City              11683
County                0
State                 0
ZipCode           48637
dtype: int64

In [18]:
df.dtypes

EventId            object
Type               object
Severity           object
StartTime(UTC)     object
EndTime(UTC)       object
TimeZone           object
AirportCode        object
LocationLat       float64
LocationLng       float64
City               object
County             object
State              object
ZipCode           float64
dtype: object

#### Convert to datetime format for StartTime(UTC) and EndTime(UTC)

In [22]:
df['StartTime(UTC)'] = pd.to_datetime(df['StartTime(UTC)']) 
df['EndTime(UTC)'] = pd.to_datetime(df['EndTime(UTC)'])

In [23]:
df.dtypes

EventId                   object
Type                      object
Severity                  object
StartTime(UTC)    datetime64[ns]
EndTime(UTC)      datetime64[ns]
TimeZone                  object
AirportCode               object
LocationLat              float64
LocationLng              float64
City                      object
County                    object
State                     object
ZipCode                  float64
dtype: object

In [59]:
df.head()

EventId  Type Severity      StartTime(UTC)        EndTime(UTC)     TimeZone  \
0     W-1  Snow    Light 2016-01-06 23:14:00 2016-01-07 00:34:00  US/Mountain   
1     W-2  Snow    Light 2016-01-07 04:14:00 2016-01-07 04:54:00  US/Mountain   
2     W-3  Snow    Light 2016-01-07 05:54:00 2016-01-07 15:34:00  US/Mountain   
3     W-4  Snow    Light 2016-01-08 05:34:00 2016-01-08 05:54:00  US/Mountain   
4     W-5  Snow    Light 2016-01-08 13:54:00 2016-01-08 15:54:00  US/Mountain   

  AirportCode  LocationLat  LocationLng      City    County State  ZipCode  
0        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
1        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
2        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
3        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0  
4        K04V      38.0972    -106.1689  Saguache  Saguache    CO  81149.0

#### Select for StartTimes(UTC) only in 2020

In [106]:
df_2020 = df.loc[(df['StartTime(UTC)'] >= '2020-01-01') & (df['StartTime(UTC)'] <= '2020-12-31')]

In [107]:
df_2020.head()

EventId  Type Severity      StartTime(UTC)        EndTime(UTC)  \
1721  W-1722  Snow    Light 2020-01-01 19:30:00 2020-01-01 22:12:00   
1722  W-1723  Snow    Light 2020-01-02 00:32:00 2020-01-02 00:50:00   
1723  W-1724  Snow    Light 2020-01-02 01:12:00 2020-01-02 01:30:00   
1724  W-1725  Snow    Light 2020-01-02 04:12:00 2020-01-02 04:30:00   
1725  W-1726  Snow    Light 2020-01-02 05:30:00 2020-01-02 06:12:00   

         TimeZone AirportCode  LocationLat  LocationLng      City    County  \
1721  US/Mountain        K04V      38.0972    -106.1689  Saguache  Saguache   
1722  US/Mountain        K04V      38.0972    -106.1689  Saguache  Saguache   
1723  US/Mountain        K04V      38.0972    -106.1689  Saguache  Saguache   
1724  US/Mountain        K04V      38.0972    -106.1689  Saguache  Saguache   
1725  US/Mountain        K04V      38.0972    -106.1689  Saguache  Saguache   

     State  ZipCode  
1721    CO  81149.0  
1722    CO  81149.0  
1723    CO  81149.0  
1724    CO  81149.0  
1725    CO  81149.0

In [108]:
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1207900 entries, 1721 to 6274205
Data columns (total 13 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   EventId         1207900 non-null  object        
 1   Type            1207900 non-null  object        
 2   Severity        1207900 non-null  object        
 3   StartTime(UTC)  1207900 non-null  datetime64[ns]
 4   EndTime(UTC)    1207900 non-null  datetime64[ns]
 5   TimeZone        1207900 non-null  object        
 6   AirportCode     1207900 non-null  object        
 7   LocationLat     1207900 non-null  float64       
 8   LocationLng     1207900 non-null  float64       
 9   City            1205311 non-null  object        
 10  County          1207900 non-null  object        
 11  State           1207900 non-null  object        
 12  ZipCode         1197991 non-null  float64       
dtypes: datetime64[ns](2), float64(3), object(8)
memory usage: 129.0+ MB


#### Check for Nulls in 2020

source: https://datatofish.com/rows-with-nan-pandas-dataframe/

In [109]:
df_2020.isnull().sum()

EventId              0
Type                 0
Severity             0
StartTime(UTC)       0
EndTime(UTC)         0
TimeZone             0
AirportCode          0
LocationLat          0
LocationLng          0
City              2589
County               0
State                0
ZipCode           9909
dtype: int64

#### Show rows with null values of any (City, Zipcode) Columns

In [110]:
df_2020_null_any = df_2020[df_2020.isnull().any(axis = 1)]
df_2020_null_any

EventId  Type  Severity      StartTime(UTC)        EndTime(UTC)  \
41087      W-41103   Fog    Severe 2020-01-04 01:35:00 2020-01-04 02:37:00   
41088      W-41104   Fog    Severe 2020-01-04 02:55:00 2020-01-04 04:05:00   
41089      W-41105  Rain     Light 2020-01-04 10:28:00 2020-01-04 10:37:00   
41090      W-41106  Rain  Moderate 2020-01-04 10:37:00 2020-01-04 10:58:00   
41091      W-41107   Fog    Severe 2020-01-04 12:42:00 2020-01-04 12:58:00   
...            ...   ...       ...                 ...                 ...   
6221999  W-6223930   Fog    Severe 2020-12-30 17:28:00 2020-12-30 17:46:00   
6222000  W-6223931  Snow     Light 2020-12-30 17:53:00 2020-12-30 18:02:00   
6222001  W-6223932  Snow  Moderate 2020-12-30 18:02:00 2020-12-30 19:53:00   
6222002  W-6223933  Snow     Light 2020-12-30 19:53:00 2020-12-30 22:53:00   
6222003  W-6223934   Fog  Moderate 2020-12-30 22:53:00 2020-12-30 23:28:00   

           TimeZone AirportCode  LocationLat  LocationLng          City  \
41087    US/Eastern        KNBT      35.0222     -76.4625         Davis   
41088    US/Eastern        KNBT      35.0222     -76.4625         Davis   
41089    US/Eastern        KNBT      35.0222     -76.4625         Davis   
41090    US/Eastern        KNBT      35.0222     -76.4625         Davis   
41091    US/Eastern        KNBT      35.0222     -76.4625         Davis   
...             ...         ...          ...          ...           ...   
6221999  US/Pacific        KHMS      46.5667    -119.6000  Hanford Site   
6222000  US/Pacific        KHMS      46.5667    -119.6000  Hanford Site   
6222001  US/Pacific        KHMS      46.5667    -119.6000  Hanford Site   
6222002  US/Pacific        KHMS      46.5667    -119.6000  Hanford Site   
6222003  US/Pacific        KHMS      46.5667    -119.6000  Hanford Site   

           County State  ZipCode  
41087    Carteret    NC      NaN  
41088    Carteret    NC      NaN  
41089    Carteret    NC      NaN  
41090    Carteret    NC      NaN  
41091    Carteret    NC      NaN  
...           ...   ...      ...  
6221999    Benton    WA      NaN  
6222000    Benton    WA      NaN  
6222001    Benton    WA      NaN  
6222002    Benton    WA      NaN  
6222003    Benton    WA      NaN  

[9909 rows x 13 columns]

In [111]:
df_2020_null_any.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9909 entries, 41087 to 6222003
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   EventId         9909 non-null   object        
 1   Type            9909 non-null   object        
 2   Severity        9909 non-null   object        
 3   StartTime(UTC)  9909 non-null   datetime64[ns]
 4   EndTime(UTC)    9909 non-null   datetime64[ns]
 5   TimeZone        9909 non-null   object        
 6   AirportCode     9909 non-null   object        
 7   LocationLat     9909 non-null   float64       
 8   LocationLng     9909 non-null   float64       
 9   City            7320 non-null   object        
 10  County          9909 non-null   object        
 11  State           9909 non-null   object        
 12  ZipCode         0 non-null      float64       
dtypes: datetime64[ns](2), float64(3), object(8)
memory usage: 1.1+ MB


#### Show rows with null values of the 'City' column

In [112]:
df_2020[df_2020['City'].isna()]

EventId  Type  Severity      StartTime(UTC)        EndTime(UTC)  \
164360    W-164392   Fog  Moderate 2020-01-01 22:53:00 2020-01-01 23:53:00   
164361    W-164393  Rain     Light 2020-01-08 09:53:00 2020-01-08 10:53:00   
164362    W-164394  Rain     Light 2020-01-09 02:53:00 2020-01-09 03:53:00   
164363    W-164395  Rain     Light 2020-01-09 13:53:00 2020-01-09 14:31:00   
164364    W-164396  Rain     Light 2020-01-09 19:53:00 2020-01-09 20:53:00   
...            ...   ...       ...                 ...                 ...   
3697160  W-3698335  Rain     Light 2020-12-24 15:38:00 2020-12-24 16:58:00   
3697161  W-3698336  Rain  Moderate 2020-12-24 16:58:00 2020-12-24 17:13:00   
3697162  W-3698337  Rain     Light 2020-12-24 17:13:00 2020-12-24 19:22:00   
3697163  W-3698338  Rain     Light 2020-12-24 20:12:00 2020-12-24 20:27:00   
3697164  W-3698339  Rain     Light 2020-12-30 22:10:00 2020-12-30 22:32:00   

           TimeZone AirportCode  LocationLat  LocationLng City  \
164360   US/Pacific        KNSI      33.2338    -119.4559  NaN   
164361   US/Pacific        KNSI      33.2338    -119.4559  NaN   
164362   US/Pacific        KNSI      33.2338    -119.4559  NaN   
164363   US/Pacific        KNSI      33.2338    -119.4559  NaN   
164364   US/Pacific        KNSI      33.2338    -119.4559  NaN   
...             ...         ...          ...          ...  ...   
3697160  US/Central        KHEY      31.3460     -85.6543  NaN   
3697161  US/Central        KHEY      31.3460     -85.6543  NaN   
3697162  US/Central        KHEY      31.3460     -85.6543  NaN   
3697163  US/Central        KHEY      31.3460     -85.6543  NaN   
3697164  US/Central        KHEY      31.3460     -85.6543  NaN   

                 County State  ZipCode  
164360   Ventura County    CA      NaN  
164361   Ventura County    CA      NaN  
164362   Ventura County    CA      NaN  
164363   Ventura County    CA      NaN  
164364   Ventura County    CA      NaN  
...                 ...   ...      ...  
3697160     Dale County    AL      NaN  
3697161     Dale County    AL      NaN  
3697162     Dale County    AL      NaN  
3697163     Dale County    AL      NaN  
3697164     Dale County    AL      NaN  

[2589 rows x 13 columns]

#### Check for nulls after dropping nulls from City and Zipcode

In [113]:
df_2020_C_Z_dropped = df_2020.dropna(how = 'any', subset = ['City', 'ZipCode'])
df_2020_C_Z_dropped.isnull().sum()

EventId           0
Type              0
Severity          0
StartTime(UTC)    0
EndTime(UTC)      0
TimeZone          0
AirportCode       0
LocationLat       0
LocationLng       0
City              0
County            0
State             0
ZipCode           0
dtype: int64

#### New Length of 2020 dropped nulls

In [114]:
len(df_2020_C_Z_dropped)

1197991

#### Null Summation
2020 original length: 1207612 

2020 rows with null values: 9908 = 1197704 *

length of dropped nulls: 1197704

**Column Zipcode nulls include all Column City nulls*

#### Summation Calculation below

In [67]:
1197704 + 9908

1207612

#### What the Summation Calculation means...
There are 9908 rows with null values and there are 9908 rows with null ZipCode values. Therefore of those 9908 rows with null Zipcode Values, there are 2589 rows with null City Values. The remainder (9908 - 2589 = 7319) rows with non-null city values and null Zipcode Values can be filled in based on Matching Latitudes and Longitudes.

Some further notes:
Googling the lat and long of entry number 164360 (33.2338	-119.4559) which is San Nicolas Island off of CA. AFAIK, San Nicolas is used as a weapons munition testing site and has neither an official ZipCode nor a 'City'. 

For the sake of this POC, we'll be dropping all rows with null values. 

#### Check for nulls of only City Column

In [115]:
df_2020['City'].isnull().sum()

2589

In [116]:
df_2020_cleaned = df_2020_C_Z_dropped.reset_index()
df_2020_cleaned = df_2020_cleaned.drop(columns = ['index'])
df_2020_cleaned

EventId  Type  Severity      StartTime(UTC)        EndTime(UTC)  \
0           W-1722  Snow     Light 2020-01-01 19:30:00 2020-01-01 22:12:00   
1           W-1723  Snow     Light 2020-01-02 00:32:00 2020-01-02 00:50:00   
2           W-1724  Snow     Light 2020-01-02 01:12:00 2020-01-02 01:30:00   
3           W-1725  Snow     Light 2020-01-02 04:12:00 2020-01-02 04:30:00   
4           W-1726  Snow     Light 2020-01-02 05:30:00 2020-01-02 06:12:00   
...            ...   ...       ...                 ...                 ...   
1197986  W-6276144  Snow     Light 2020-12-14 17:30:00 2020-12-15 01:53:00   
1197987  W-6276145  Snow  Moderate 2020-12-15 01:53:00 2020-12-15 03:19:00   
1197988  W-6276146  Snow     Light 2020-12-15 03:19:00 2020-12-15 05:53:00   
1197989  W-6276147  Snow     Light 2020-12-23 04:24:00 2020-12-23 05:53:00   
1197990  W-6276148  Snow     Light 2020-12-23 10:53:00 2020-12-23 11:26:00   

            TimeZone AirportCode  LocationLat  LocationLng      City  \
0        US/Mountain        K04V      38.0972    -106.1689  Saguache   
1        US/Mountain        K04V      38.0972    -106.1689  Saguache   
2        US/Mountain        K04V      38.0972    -106.1689  Saguache   
3        US/Mountain        K04V      38.0972    -106.1689  Saguache   
4        US/Mountain        K04V      38.0972    -106.1689  Saguache   
...              ...         ...          ...          ...       ...   
1197986  US/Mountain        KBVR      42.5833    -108.2833    Lander   
1197987  US/Mountain        KBVR      42.5833    -108.2833    Lander   
1197988  US/Mountain        KBVR      42.5833    -108.2833    Lander   
1197989  US/Mountain        KBVR      42.5833    -108.2833    Lander   
1197990  US/Mountain        KBVR      42.5833    -108.2833    Lander   

           County State  ZipCode  
0        Saguache    CO  81149.0  
1        Saguache    CO  81149.0  
2        Saguache    CO  81149.0  
3        Saguache    CO  81149.0  
4        Saguache    CO  81149.0  
...           ...   ...      ...  
1197986   Fremont    WY  82520.0  
1197987   Fremont    WY  82520.0  
1197988   Fremont    WY  82520.0  
1197989   Fremont    WY  82520.0  
1197990   Fremont    WY  82520.0  

[1197991 rows x 13 columns]

In [117]:
df_2020_cleaned.to_csv('weather_2020.csv', index = True)

#### Get data from Jan 2020 to April 2020

In [118]:
df_jan_apr = df_2020_cleaned.loc[
            (df_2020_cleaned['StartTime(UTC)'] >= '2020-01-01') & 
            (df_2020_cleaned['StartTime(UTC)'] <= '2020-04-30')]
df_jan_apr = df_jan_apr.reset_index().drop(columns=['index'])

df_jan_apr.to_csv('weather_jan_apr_2020.csv', index = True)

#### Get data from Jan 2020 to August 2020

In [121]:
df_2020_jan_aug = df_2020_cleaned.loc[
                (df_2020_cleaned['StartTime(UTC)'] >= '2020-01-01') &
                (df_2020_cleaned['StartTime(UTC)'] <= '2020-08-31')
].reset_index().drop(columns=['index'])
df_2020_jan_aug.to_csv('weather_jan_aug_2020.csv', index = True)